README.md
# [mod_autoindex](https://github.com/hex7c0/mod_autoindex)

[![NPM version](https://img.shields.io/npm/v/mod_autoindex.svg)](https://www.npmjs.com/package/mod_autoindex)
[![Linux Status](https://img.shields.io/travis/hex7c0/mod_autoindex.svg?label=linux)](https://travis-ci.org/hex7c0/mod_autoindex)
[![Windows Status](https://img.shields.io/appveyor/ci/hex7c0/mod_autoindex.svg?label=windows)](https://ci.appveyor.com/project/hex7c0/mod_autoindex)
[![Dependency Status](https://img.shields.io/david/hex7c0/mod_autoindex.svg)](https://david-dm.org/hex7c0/mod_autoindex)
[![Coveralls](https://img.shields.io/coveralls/hex7c0/mod_autoindex.svg)](https://coveralls.io/r/hex7c0/mod_autoindex)

Generates directory indexes, automatically, similar to the Unix `ls` command or the Win32 `dir` shell command; with memoization.
In combination with [serve-static](https://github.com/expressjs/serve-static).

Display index of given root directory, like https://code.angularjs.org/2.0.0-alpha.30/ or `apache2 mod_autoindex`

## Installation

Install through NPM

```bash
npm install mod_autoindex
```
or
```bash
git clone git://github.com/hex7c0/mod_autoindex.git
```

## API

inside expressjs project
```js
var autoindex = require('mod_autoindex');

var app = require('express')();

app.use(autoindex(__dirname));
```

### autoindex(root [, options])

#### root

 - `root` - **String** Index given root directory *(default "required")*

#### [options]

 - `exclude` - **RegExp** Regular expression for files/dirs exclude *(default "disabled")*
 - `dotfiles`- **Boolean** Flag for hide dotfiles *(default "true")*
 - `date` - **Boolean** Flag for display modification time *(default "true")*
 - `size` - **Boolean** Flag for display files size *(default "true")*
 - `priority` - **Boolean** Flag for display dirs before files *(default "true")*
 - `cache` - **Boolean** Flag for using cache (depends from mtime dir) *(default "true")*
 - `strictMethod` - **Boolean** Flag for show "HEAD" and "GET" HTTP methods only *(default "false")*
 - `sync` - **Boolean** Flag for using "sync" methods instead of callback *(default "false")*
 - `json` - **Boolean** Flag for display json output instead of html *(default "false")*
 - `static` - **Object | false** Options for [serve-static](https://github.com/expressjs/serve-static) or disable support (if you use static server like nginx) *(default "{}")*

## Examples

Take a look at my [examples](examples)

### [License GPLv3](LICENSE)


History.md
v1.5.4 / 2015-10-05
==================

  * Tested with node@4

v1.5.3 / 2015-08-22
==================

  * Tested with iojs@3

v1.5.2 / 2015-06-18
==================

  * Update `serve-static`@1.10.0

v1.5.1 / 2015-05-19
==================

  * Tested with iojs@2
  * SPDX license
  * Update `serve-static`@1.9.3

v1.5.0 / 2015-03-16
==================

  * Update `serve-static`@1.9.2
  * Fix serve-static on exclude files
  * `windows` test
  * `iojs` test
  * `coveralls` test

v1.4.19 / 2015-01-26
==================

  * Package fix
  * Update `serve-static`@1.8.1

v1.4.18 / 2015-01-06
==================

  * Update `serve-static`@1.8.0

v1.4.17 / 2015-01-01
==================

  * Update devDependencies

v1.4.16 / 2014-11-16
==================

  * Fix web Status Code
  * Strict check

v1.4.15 / 2014-10-27
==================

  * `jshint`

v1.4.14 / 2014-10-24
==================

  * Update `server-static`@1.7.1

v1.4.13 / 2014-10-15
==================

  * Update `server-static`@1.7.0

v1.4.12 / 2014-10-09
==================

  * Update `server-static`@1.6.4

v1.4.11 / 2014-10-04
==================

  * Remove jsdoc
  * `package.json` min

v1.4.10 / 2014-09-26
==================

  * Update `server-static`@1.6.3

v1.4.9 / 2014-09-18
==================

  * Update `server-static`@1.6.2

v1.4.8 / 2014-09-10
==================

  * Update `server-static`@1.6.1

v1.4.7 / 2014-09-05
==================

  * Update `server-static`@1.5.3

v1.4.5 / 2014-08-24
==================

  * Sync ignore fix

v1.4.4 / 2014-08-23
==================

  * Performance tips
  * Update `server-static`@1.5.3

v1.4.3 / 2014-08-16
==================

  * Fix critical issue
  * Update `server-static`@1.5.2

v1.4.1 / 2014-08-11
==================

  * Update README.md

1.4.0 / 2014-08-11
==================

  * Add "sync" (options)
  * Add "json" (options)
  * Add "cache" (options)
  * Full callback work for "callback" option
  * Using `parseurl`@1.3.0 instead of `url.parse`

1.3.1 / 2014-08-08
==================

  * Add "dotfiles" (options)
  * Chance to disable `serve-static` (if you use static server like nginx)

1.2.1 / 2014-08-07
==================

  * Change parsing or "req.url"

1.2.0 / 2014-08-06
==================

  * Add "static" (options)
  * `serve-static` implemented

1.1.0 / 2014-08-05
==================

  * Add options
  * Show directories first
  * Mocha testing

1.0.0 / 2014-08-04
==================

  * Project start


test/sync.js
'use strict';
/**
 * @file sync test
 * @module mod_autoindex
 * @subpackage test
 * @version 0.0.1
 * @author hex7c0 <hex7c0@gmail.com>
 * @license GPLv3
 */

/*
 * initialize module
 */
var index = require('..');
var express = require('express');
var request = require('supertest');
var assert = require('assert');

/*
 * test module
 */
describe('sync', function() {

  var app, text;

  beforeEach(function(done) {

    app = express();
    done();
  });

  describe('found', function() {

    it('should return 200 callback mod_index', function(done) {

      app.use(index(__dirname, {
        json: true
      }));

      request(app).get('/').expect(200).end(function(err, res) {

        assert.equal(err, null);
        text = JSON.parse(res.text);
        assert.equal(typeof text, 'object');
        done();
      });
    });
    it('should return 200 sync mod_index', function(done) {

      app.use(index(__dirname, {
        sync: true,
        json: true
      }));

      request(app).get('/').expect(200).end(function(err, res) {

        assert.equal(err, null);
        assert.deepEqual(text, JSON.parse(res.text));
        done();
      });
    });
  });

  describe('not found', function() {

    it('should return 400 callback mod_index', function(done) {

      app.use(index(__dirname, {
        json: true
      })).use(function(err, req, res, next) {

        var code;
        switch (err.message.toLowerCase()) {
          case 'forbidden':
            code = 403;
            break;
          case 'not found':
            code = 404;
            break;
          case 'request-uri too large':
            code = 414;
            break;
          default:
            code = 500;
            break;
        }
        res.status(code).send('error');
      });

      request(app).get('/foo').expect(404).end(function(err, res) {

        assert.equal(err, null);
        assert.notDeepEqual(res.text, text);
        text = res.text;
        assert.equal(typeof text, 'string');
        done();
      });
    });
    it('should return 500 sync mod_index', function(done) {

      app.use(index(__dirname, {
        sync: true,
        json: true
      })).use(function(err, req, res, next) {

        var code;
        switch (err.message.toLowerCase()) {
          case 'forbidden':
            code = 403;
            break;
          case 'not found':
            code = 404;
            break;
          case 'request-uri too large':
            code = 414;
            break;
          default:
            code = 500;
            break;
        }
        res.status(code).send('error');
      });

      request(app).get('/foo').expect(500).end(function(err, res) {

        assert.equal(err, null);
        assert.equal(text, res.text);
        done();
      });
    });
  });
});


test/basic.js
'use strict';
/**
 * @file basic test
 * @module mod_autoindex
 * @subpackage test
 * @version 0.0.1
 * @author https://github.com/expressjs/serve-static & https://github.com/expressjs/serve-index
 * @license GPLv3
 */

/*
 * initialize module
 */
var index = require('..');
var app = require('express')();
var request = require('supertest');

/*
 * test module
 */
describe('basic', function() {

  var dir = __dirname + '/..'; // project dir

  describe('callback', function() {

    before(function(done) {

      app.use(index(dir)).use(function(err, req, res, next) {

        var code = 0;
        switch (err.message.toLowerCase()) {
          case 'bad request':
            code = 400;
            break;
          case 'forbidden':
            code = 403;
            break;
          case 'not found':
            code = 404;
            break;
          case 'request-uri too large': // process.version < iojs@3
          case 'uri too long': // process.version >= iojs@3
            code = 414;
            break;
          default:
            code = 500;
            break;
        }
        res.status(code).send('error');
        return;
      });
      done();
    });

    it('should ignore hidden files', function(done) {

      request(app).get('/.npmignore').expect(404, done);
    });
    it('should serve static files', function(done) {

      request(app).get('/index.js').expect(200, done);
    });
    it('should support nesting', function(done) {

      request(app).get('/test/basic.js').expect(200, done);
    });
    it('should support multiple slahes', function(done) {

      request(app).get('//////examples').expect(200, done);
    });
    it('should set Content-Type', function(done) {

      request(app).get('/index.js').expect('Content-Type',
        'application/javascript').expect(200, done);
    });
    it('should not support urlencoded pathnames', function(done) {

      request(app).get('/foo%20bar').expect(404, done);
    });
    it('should not choke on auth-looking URL', function(done) {

      request(app).get('//todo@txt').expect(404, done);
    });
    it('should support index.html', function(done) {

      request(app).get('/test').expect('Content-Type', /html/)
          .expect(200, done);
    });
    it('should not support if dir not exist', function(done) {

      request(app).get('/tests').expect(404, done);
    });
    it('should support ../', function(done) {

      request(app).get('/test/../index.js').expect(200, done);
    });
    it('should support HEAD', function(done) {

      request(app).head('/index.js').expect(200, '', done);
    });
    it('should support conditional requests', function(done) {

      request(app).get('/index.js').end(
        function(err, res) {

          if (err) throw err;
          request(app).get('/index.js').set('If-None-Match', res.headers.etag)
              .expect(304, done);
        });
    });
    it('should set max-age=0 by default', function(done) {

      request(app).get('/index.js')
          .expect('cache-control', 'public, max-age=0').expect(200, done);
    });
    it('should serve text/html without Accept header', function(done) {

      request(app).get('/').expect('Content-Type', 'text/html; charset=utf-8')
          .expect(200, done);
    });
    it('should deny path will NULL byte', function(done) {

      request(app).get('/%00').expect(400, done);
    });
    it('should deny path outside root', function(done) {

      request(app).get('/../').expect(403, done);
    });
    it('should deny directory traversal attack', function(done) {

      request(app).get('/../../../../../../../../../../../').expect(403, done);
    });
    it('should deny directory traversal attack', function(done) {

      request(app).get('/~/Workspace').expect(404, done);
    });
    it('should treat an ENAMETOOLONG as a 414', function(done) {

      var path = Array(11000).join('foobar');
      request(app).get('/' + path).expect(414, done);
    });
  });

  describe('sync', function() {

    before(function(done) {

      app.use(index(dir, {
        sync: true
      }));
      app.use(function(err, req, res, next) {

        var code = 0;
        switch (err.message.toLowerCase()) {
          case 'bad request':
            code = 400;
            break;
          case 'forbidden':
            code = 403;
            break;
          case 'not found':
            code = 404;
            break;
          case 'request-uri too large':
            code = 414;
            break;
          default:
            code = 500;
            break;
        }
        res.status(code).send('error');
        return;
      });
      done();
    });

    it('should ignore hidden files', function(done) {

      request(app).get('/.npmignore').expect(404, done);
    });
    it('should serve static files', function(done) {

      request(app).get('/index.js').expect(200, done);
    });
    it('should support nesting', function(done) {

      request(app).get('/test/basic.js').expect(200, done);
    });
    it('should support multiple slahes', function(done) {

      request(app).get('//////examples').expect(200, done);
    });
    it('should set Content-Type', function(done) {

      request(app).get('/index.js').expect('Content-Type',
        'application/javascript').expect(200, done);
    });
    it('should not support urlencoded pathnames', function(done) {

      request(app).get('/foo%20bar').expect(404, done);
    });
    it('should not choke on auth-looking URL', function(done) {

      request(app).get('//todo@txt').expect(404, done);
    });
    it('should support index.html', function(done) {

      request(app).get('/test').expect('Content-Type', /html/)
          .expect(200, done);
    });
    it('should not support if dir not exist', function(done) {

      request(app).get('/tests').expect(404, done);
    });
    it('should support ../', function(done) {

      request(app).get('/test/../index.js').expect(200, done);
    });
    it('should support HEAD', function(done) {

      request(app).head('/index.js').expect(200, '', done);
    });
    it('should support conditional requests', function(done) {

      request(app).get('/index.js').end(
        function(err, res) {

          if (err) throw err;
          request(app).get('/index.js').set('If-None-Match', res.headers.etag)
              .expect(304, done);
        });
    });
    it('should set max-age=0 by default', function(done) {

      request(app).get('/index.js')
          .expect('cache-control', 'public, max-age=0').expect(200, done);
    });
    it('should serve text/html without Accept header', function(done) {

      request(app).get('/').expect('Content-Type', 'text/html; charset=utf-8')
          .expect(200, done);
    });
    it('should deny path will NULL byte', function(done) {

      request(app).get('/%00').expect(400, done);
    });
    it('should deny path outside root', function(done) {

      request(app).get('/../').expect(403, done);
    });
    it('should deny directory traversal attack', function(done) {

      request(app).get('/../../../../../../../../../../../').expect(403, done);
    });
    it('should deny directory traversal attack', function(done) {

      request(app).get('/~/Workspace').expect(404, done);
    });
    it('should treat an ENAMETOOLONG as a 414', function(done) {

      var path = Array(11000).join('foobar');
      request(app).get('/' + path).expect(414, done);
    });
  });

  describe('dotfiles', function() {

    describe('callback', function() {

      before(function(done) {

        app.use(index(dir, {
          dotfiles: false,
        }));
        app.use(function(err, req, res, next) {

          var code = 0;
          switch (err.message.toLowerCase()) {
            case 'bad request':
              code = 400;
              break;
            case 'forbidden':
              code = 403;
              break;
            case 'not found':
              code = 404;
              break;
            case 'request-uri too large':
              code = 414;
              break;
            default:
              code = 500;
              break;
          }
          res.status(code).send('error');
          return;
        });
        done();
      });

      it('shouldn\'t ignore hidden files', function(done) {

        request(app).get('/.npmignore').expect(200, done);
      });
      it('should serve static files', function(done) {

        request(app).get('/index.js').expect(200, done);
      });
      it('should support nesting', function(done) {

        request(app).get('/test/basic.js').expect(200, done);
      });
      it('should support multiple slahes', function(done) {

        request(app).get('//////examples').expect(200, done);
      });
      it('should set Content-Type', function(done) {

        request(app).get('/index.js').expect('Content-Type',
          'application/javascript').expect(200, done);
      });
      it('should not support urlencoded pathnames', function(done) {

        request(app).get('/foo%20bar').expect(404, done);
      });
      it('should not choke on auth-looking URL', function(done) {

        request(app).get('//todo@txt').expect(404, done);
      });
      it('should support index.html', function(done) {

        request(app).get('/test').expect('Content-Type', /html/).expect(200,
          done);
      });
      it('should not support if dir not exist', function(done) {

        request(app).get('/tests').expect(404, done);
      });
      it('should support ../', function(done) {

        request(app).get('/test/../index.js').expect(200, done);
      });
      it('should support HEAD', function(done) {

        request(app).head('/index.js').expect(200, '', done);
      });
      it('should support conditional requests', function(done) {

        request(app).get('/index.js').end(
          function(err, res) {

            if (err) throw err;
            request(app).get('/index.js')
                .set('If-None-Match', res.headers.etag).expect(304, done);
          });
      });
      it('should set max-age=0 by default', function(done) {

        request(app).get('/index.js').expect('cache-control',
          'public, max-age=0').expect(200, done);
      });
      it('should serve text/html without Accept header', function(done) {

        request(app).get('/')
            .expect('Content-Type', 'text/html; charset=utf-8').expect(200,
              done);
      });
      it('should deny path will NULL byte', function(done) {

        request(app).get('/%00').expect(400, done);
      });
      it('should deny path outside root', function(done) {

        request(app).get('/../').expect(403, done);
      });
      it('should deny directory traversal attack', function(done) {

        request(app).get('/../../../../../../../../../../../')
            .expect(403, done);
      });
      it('should deny directory traversal attack', function(done) {

        request(app).get('/~/Workspace').expect(404, done);
      });
      it('should treat an ENAMETOOLONG as a 414', function(done) {

        var path = Array(11000).join('foobar');
        request(app).get('/' + path).expect(414, done);
      });
    });

    describe('sync', function() {

      before(function(done) {

        app.use(index(dir, {
          dotfiles: false,
          sync: true
        }));
        app.use(function(err, req, res, next) {

          var code = 0;
          switch (err.message.toLowerCase()) {
            case 'bad request':
              code = 400;
              break;
            case 'forbidden':
              code = 403;
              break;
            case 'not found':
              code = 404;
              break;
            case 'request-uri too large':
              code = 414;
              break;
            default:
              code = 500;
              break;
          }
          res.status(code).send('error');
          return;
        });
        done();
      });

      it('shouldn\'t ignore hidden files', function(done) {

        request(app).get('/.npmignore').expect(200, done);
      });
      it('should serve static files', function(done) {

        request(app).get('/index.js').expect(200, done);
      });
      it('should support nesting', function(done) {

        request(app).get('/test/basic.js').expect(200, done);
      });
      it('should support multiple slahes', function(done) {

        request(app).get('//////examples').expect(200, done);
      });
      it('should set Content-Type', function(done) {

        request(app).get('/index.js').expect('Content-Type',
          'application/javascript').expect(200, done);
      });
      it('should not support urlencoded pathnames', function(done) {

        request(app).get('/foo%20bar').expect(404, done);
      });
      it('should not choke on auth-looking URL', function(done) {

        request(app).get('//todo@txt').expect(404, done);
      });
      it('should support index.html', function(done) {

        request(app).get('/test').expect('Content-Type', /html/).expect(200,
          done);
      });
      it('should not support if dir not exist', function(done) {

        request(app).get('/tests').expect(404, done);
      });
      it('should support ../', function(done) {

        request(app).get('/test/../index.js').expect(200, done);
      });
      it('should support HEAD', function(done) {

        request(app).head('/index.js').expect(200, '', done);
      });
      it('should support conditional requests', function(done) {

        request(app).get('/index.js').end(
          function(err, res) {

            if (err) throw err;
            request(app).get('/index.js')
                .set('If-None-Match', res.headers.etag).expect(304, done);
          });
      });
      it('should set max-age=0 by default', function(done) {

        request(app).get('/index.js').expect('cache-control',
          'public, max-age=0').expect(200, done);
      });
      it('should serve text/html without Accept header', function(done) {

        request(app).get('/')
            .expect('Content-Type', 'text/html; charset=utf-8').expect(200,
              done);
      });
      it('should deny path will NULL byte', function(done) {

        request(app).get('/%00').expect(400, done);
      });
      it('should deny path outside root', function(done) {

        request(app).get('/../').expect(403, done);
      });
      it('should deny directory traversal attack', function(done) {

        request(app).get('/../../../../../../../../../../../')
            .expect(403, done);
      });
      it('should deny directory traversal attack', function(done) {

        request(app).get('/~/Workspace').expect(404, done);
      });
      it('should treat an ENAMETOOLONG as a 414', function(done) {

        var path = Array(11000).join('foobar');
        request(app).get('/' + path).expect(414, done);
      });
    });
  });
});


Gruntfile.js
'use strict';
/**
 * @file gruntfile
 * @subpackage main
 * @version 0.0.1
 * @author hex7c0 <hex7c0@gmail.com>
 * @copyright hex7c0 2014
 * @license GPLv3
 */

module.exports = function(grunt) {

    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        banner: '/*\n' + ' * <%= pkg.name %> v<%= pkg.version %>\n'
                + ' * (c) <%= pkg.author.name %> <%= pkg.homepage %>\n'
                + ' * Licensed under <%= pkg.license %>\n' + ' */\n',

        clean: [ 'index.min.js', 'min/**/*.js' ],

        uglify: {
            options: {
                preserveComments: 'false'
            },
            target: {
                files: [ {
                    expand: true,
                    src: 'lib/**/*.js',
                    dest: 'min'
                }, {
                    expand: true,
                    src: 'module/**/*.js',
                    dest: 'min'
                }, {
                    'index.min.js': 'index.js'
                } ]
            }
        },

        shell: {
            options: {
                failOnError: false
            },
            docs: {
                command: 'jsdoc ./lib/*.js ./module/*.js -c .jsdoc.json'
            }
        },

        endline: {
            target: {
                options: {
                    except: 'node_modules',
                    replaced: true
                },
                files: [ {
                    src: './**/*.js'
                } ]
            }
        }
    });

    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-endline');

    grunt.registerTask('min', [ 'clean', 'uglify', 'endline' ]);

    return;
};


examples/error.js
'use strict';
/**
 * @file error example
 * @module mod_autoindex
 * @subpackage examples
 * @version 0.0.1
 * @author hex7c0 <hex7c0@gmail.com>
 * @license GPLv3
 */

/*
 * initialize module
 */
var index = require('..'); // use require('mod_autoindex') instead
var express = require('express');

var dir = __dirname + '/..';
var app = express();

// index
app.use(index(dir));

// error handler
app.use(function(err, req, res, next) {

  var code;
  var msg = err.message.toLowerCase();

  switch (msg) {
    case 'forbidden':
      code = 403;
      break;
    case 'not found':
      code = 404;
      break;
    case 'request-uri too large': // process.version < iojs@3
    case 'uri too long': // process.version >= iojs@3
      code = 414;
      break;
    default:
      code = 500;
      break;
  }
  res.status(code).send(msg);
}).listen(3000);// server starting
console.log('starting "hello world" on port 3000');


index.js
'use strict';
/**
 * @file mod_autoindex main
 * @module mod_autoindex
 * @subpackage main
 * @version 1.5.0
 * @author hex7c0 <hex7c0@gmail.com>
 * @copyright hex7c0 2014
 * @license GPLv3
 */

/*
 * initialize module
 */
var status = require('http').STATUS_CODES;
var path = require('path');
var fs = require('fs');
var parse = require('parseurl');
var serve = require('serve-static');
var header = '<html>\n<head><title>Index of {{path}}</title></head>\n<body bgcolor="white">\n<h1>Index of {{path}}</h1><hr><pre>\n';
var footer = '</pre><hr></body>\n</html>\n';
var month = [ 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep',
  'Oct', 'Nov', 'Dec' ];

/*
 * functions
 */
/**
 * build multiple space
 * 
 * @function multiple
 * @param {Integer} length - string.length
 * @param {Integer} limit - max space
 * @return {String}
 */
function multiple(length, limit) {

  var space = '';
  if (limit < length) {
    return space;
  }
  for (var i = 0, ii = limit - length; i < ii; ++i) {
    space += ' ';
  }
  return space;
}

/**
 * error handling
 * 
 * @function error
 * @param {Object|Number} e - error
 * @return {Error}
 */
function error(e) {

  if (isNaN(e) === false) {
    return new Error(status[e]);
  }
  return new Error(e.code === 'ENAMETOOLONG' ? status[414] : status[404]);
}

/**
 * add zero
 * 
 * @function pad
 * @param {Integer} num - number
 * @return {String}
 */
function pad(num) {

  var norm = Math.abs(Math.floor(num));
  return (norm < 10 ? '0' : '') + norm;
}

/**
 * function wrapper for multiple require
 * 
 * @function wrapper
 * @param {Object} my - options
 * @return {Object}
 */
function wrapper(my) {

  /**
   * next callback
   * 
   * @function output
   * @param {Object} res - response to client
   * @param {Object|String} head - header
   * @param {String} after - post header
   * @param {String} paths - dir path
   * @param {Object} stat - path stat
   */
  function output(res, head, after, paths, stat) {

    if (my.json === true) {
      res.send(head);
      STORY.body = head;
    } else {
      head += after + footer;
      res.send(head);
      STORY.body = head;
    }
    if (my.cache === true) {
      STORY.path = paths;
      STORY.mtime = stat.mtime.getTime();
    }
    return;
  }

  /**
   * json builder
   * 
   * @function json
   * @param {Object} head - header
   * @param {String} after - post header
   * @param {String} file - file name
   * @param {Object} stats - stats of file
   * @return {Array}
   */
  function json(head, after, file, stats) {

    var size;
    var fil = file;
    if (stats.isDirectory() === true) {
      size = '-';
      fil += '/';
    } else {
      size = String(stats.size);
    }
    head[fil] = Object.create(null);
    if (my.date === true) {
      var d = new Date(stats.mtime);
      var h = pad(d.getDate()) + '-' + month[d.getMonth()] + '-'
        + d.getFullYear();
      h += ' ' + pad(d.getHours()) + ':' + pad(d.getMinutes());
      head[fil].mtime = h;
    }
    if (my.size === true) {
      head[fil].size = size;
    }
    return [ head, null ];
  }

  /**
   * html builder
   * 
   * @function html
   * @param {String} head - header
   * @param {String} after - post header
   * @param {String} file - file name
   * @param {Object} stats - stats of file
   * @return {Array}
   */
  function html(head, after, file, stats) {

    var h;
    var size;
    var hea = head;
    var afte = after;
    var fil = file;
    if (stats.isDirectory() === true) {
      size = '-';
      fil += '/';
    } else {
      size = String(stats.size);
    }
    h = '<a href="' + fil + '">' + fil + '</a>';
    h += multiple(fil.length, 50);
    if (my.date === true) {
      var d = new Date(stats.mtime);
      h += pad(d.getDate()) + '-' + month[d.getMonth()] + '-' + d.getFullYear();
      h += ' ' + pad(d.getHours()) + ':' + pad(d.getMinutes());
      h += multiple(size.length, 20);
    }
    if (my.size === true) {
      h += size;
    }
    h += '\n';
    if (my.priority === true) {
      if (size === '-') {
        hea += h;
      } else {
        afte += h;
      }
    } else {
      hea += h;
    }
    return [ hea, afte ];
  }

  var STORY = Object.create(null);
  var build = html;
  if (my.json === true) {
    build = json;
  }

  // start sync
  if (my.sync === true) {
    /**
     * body
     * 
     * @function mod_sync
     * @param {Object} req - client request
     * @param {Object} res - response to client
     * @param {next} next - next callback
     */
    return function mod_sync(req, res, next) {

      if (my.strictMethod === true && 'GET' !== req.method
        && 'HEAD' !== req.method) {
        return next(error(404));
      }
      var paths = decodeURIComponent(parse(req).pathname)
          .replace(/\/{2,}/, '/');
      var prova = path.normalize(path.join(my.root, paths));
      if (~prova.indexOf('\0')) { // null byte(s), bad request
        return next(error(400));
      } else if (prova <= my.root) { // /../
        return next(error(403));
      }

      var stat = fs.statSync(prova);
      if (stat) {
        if (STORY.mtime && STORY.mtime === stat.mtime.getTime()
          && STORY.path === prova) { // cache
          return res.send(STORY.body);
        }
        if (stat.isDirectory() === false) {
          if (my.exclude !== false && my.exclude.test(prova)) {
            return next(error(404));
          }
          return my.statico(req, res, next);
        }
        var head;
        if (my.json === true) {
          head = Object.create(null);
        } else {
          head = header;
          head = head.replace(/{{path}}/g, paths);
          if (paths != '/') {
            head += '<a href="../">../</a>\n';
          }
        }
        var files = fs.readdirSync(prova);
        if (files) {
          var after = '';
          for (var i = 0, ii = files.length; i < ii; ++i) {

            var file = files[i];
            if (my.exclude !== false && my.exclude.test(file)) {
              continue;
            }
            if (my.dotfiles === true && file[0] === '.') {
              continue;
            }
            var stats = fs.statSync(prova + path.sep + file);
            if (stats) {
              var r = build(head, after, file, stats);
              head = r[0];
              after = r[1];
            }
          }
          return output(res, head, after, prova, stat);

        }
      }
      return my.statico(req, res, next);

    };
  }
  // end sync

  // start callback
  /**
   * body
   * 
   * @function mod_callback
   * @param {Object} req - client request
   * @param {Object} res - response to client
   * @param {next} next - next callback
   */
  return function mod_callback(req, res, next) {

    if (my.strictMethod === true && 'GET' !== req.method
      && 'HEAD' !== req.method) {
      return next(error(404));
    }
    var paths = decodeURIComponent(parse(req).pathname).replace(/\/{2,}/, '/');
    var prova = path.normalize(path.join(my.root, paths));
    if (~prova.indexOf('\0')) { // null byte(s), bad request
      return next(error(400));
    } else if (prova <= my.root) { // /../
      return next(error(403));
    }

    return fs.stat(prova, function(err, stat) {

      if (err) {
        return next(error(err));
      }
      if (STORY.mtime && STORY.mtime === stat.mtime.getTime()
        && STORY.path === prova) { // cache
        return res.send(STORY.body);
      }
      if (stat.isDirectory() === false) {
        if (my.exclude !== false && my.exclude.test(prova)) {
          return next(error(404));
        }
        return my.statico(req, res, next);
      }
      var head;
      if (my.json === true) {
        head = Object.create(null);
      } else {
        head = header;
        head = head.replace(/{{path}}/g, paths);
        if (paths != '/') {
          head += '<a href="../">../</a>\n';
        }
      }

      return fs.readdir(prova, function(err, files) {

        if (err) {
          return next(error(err));
        }

        var after = '';
        var cc = files.length - 1;
        for (var i = 0, ii = files.length; i < ii; ++i) {
          !function(file) {

            if (my.exclude !== false && my.exclude.test(file)) {
              if (cc === 0) {
                return output(res, head, after, prova, stat);
              }
              --cc;
              return;
            } else if (my.dotfiles === true && file[0] === '.') {
              if (cc === 0) {
                return output(res, head, after, prova, stat);
              }
              --cc;
              return;
            }

            return fs.stat(prova + path.sep + file, function(err, stats) {

              if (err) {
                return next(error(err));
              }
              var r = build(head, after, file, stats);
              head = r[0];
              after = r[1];
              if (cc === 0) {
                return output(res, head, after, prova, stat);
              }
              --cc;
              return;
            });
          }(files[i]);
        }

        return;
      });
    });
  };
  // end callback

}

/**
 * options setting
 * 
 * @exports index
 * @function index
 * @param {String} root - root path
 * @param {Object} [opt] - various options. Check README.md
 * @return {Object}
 */
function index(root, opt) {

  if (!root) {
    throw new TypeError('root path required');
  }
  var r = path.resolve(root);
  if (r[r.length - 1] == '/') {
    r = r.substr(0, r.length - 1);
  }
  if (!fs.existsSync(root)) {
    throw new Error('path not exists');
  } else if (!fs.statSync(r).isDirectory()) {
    throw new Error('path is not a directory');
  }
  var options = opt || Object.create(null);
  var my = {
    root: r,
    exclude: options.exclude || false,
    dotfiles: options.dotfiles === false ? false : true,
    date: options.date === false ? false : true,
    size: options.size === false ? false : true,
    priority: options.priority === false ? false : true,
    cache: options.cache === false ? false : true,
    strictMethod: Boolean(options.strictMethod),
    sync: Boolean(options.sync),
    json: Boolean(options.json),
    static: options.static === false ? false : options.static || {}
  };

  if (my.dotfiles === false && options.static !== false) {
    my.static.dotfiles = 'allow';
  }
  my.statico = my.static === false ? function(req, res, next) {

    return next();
  } : serve(r, my.static);

  return wrapper(my);
}
module.exports = index;
